enhancements/etcd/assisted-install.md

```diff
@@ -0,0 +1,221 @@
+
+## Summary
+The assisted installer was designed to utilize three bare-metal machines to
+provision the control-plane. In order to do this after bootstrap is complete
+the bootstrap node is repurposed into the third master node.
+
+## Challenges
+The installer currently does not provide an `install_type` key for the
+install-config. So no direct signal is passed from the installer to operator of
+the exact intent to perform a non standard install process.
+
+The result is the operator must make educated assumptions as to the install intent based
+on observations from the cluster itself.
+
+## Motivation
+Currently the `cluster-etcd-operator` does not tolerate less than 3 master
+nodes during install time.
+
+Reasons outlined for cluster with less than 3 master nodes as unsafe non HA.
+1. Violates operational best practices for etcd - unstable
+1. Allows non-HA configurations which we cannot support in production - unsafe and non-HA
+1. Allows a situation where we can get stuck unable to re-achieve quorum, resulting in cluster-death - unsafe, non-HA, non-production, unstable
+1. The combination of all these things makes the situation unsupportable. 
+
+To allow toleration of this check we provided an unsupportedConfigOverrides
+`useUnsupportedUnsafeNonHANonProductionUnstableEtcd`. As assisted installer is
+now moving to GA a supported toleration is required.
+
+## Goals
+
+1.  Provide a supported solution for assisted installer that will allow
+    `cluster-etcd-operator` to tolerate 2 master replicas as install complete.
+
+## Non-Goals
+
+1.  Support the current unsupportedConfigOverrides
+1.  Add support for less than 3 master node clusters
+
+## Currently Supported Functionality 
+Currently `cluster-etcd-operator` and OpenShift only supports clusters with
+three master nodes as install complete.
+
+### EnvVarController
+Currently the `EnvVarController` ensures that all of the data required for the
+etcd static pods exists. A condition exists within the `getEtcdEnvVars`[1]  function which checks if we have the expected master node count or if
+`useUnsupportedUnsafeNonHANonProductionUnstableEtcd` is populated otherwise
+return error which results in the controller being degraded and no static-pod
+spec installed on the master node for etcd.
+
+[1] https://github.com/openshift/cluster-etcd-operator/blob/be2b8d97ca048baecb051d288971a9f6db411457/pkg/etcdenvvar/etcd_env.go#L71
+
+## Proposed changes
+The following changes are proposed.
+
+In order for the operator to tolerate an unsafe install configuration all
+efforts must be taken to minimize the risk of tolerating user defined unsafe
+cluster. For example to save on costs user chooses 2 master replicas.
+
+1.  Observe master replica count from install-config.
+1.  Observe install status and only tolerate less than 3 masters until install
+    complete. After install status is complete the operator will go into a
+degraded state until the cluster has 3 or more master nodes.
```

Comment:
perhaps this could be even more specific, e.g we expect to tolerate less than 3 masters only between bootstrap-complete and install-complete?

Reply:
how ceo is aware of it?

Reply:
@romfreiman see the " bootstrap configmap" section below, I think that explains it?

Reply:
@hexfusion 

I understand the concept behind the gating condition proposed in the "bootstrap configmap" section below to tolerate lack of quorum during the period detected as the "bootstrap phase" (which we can take separately for now) — my question is about what observable cluster state we should use to decide when to enable that toleration. 

It _seems_ like the proposal is suggesting that the user ("assisted installer" in this case) provides a master replica count of 2 through `controlPlane` field of install config which is interpreted by the platform as indicating this is a "compact" cluster (for lack of a better word) which we change behavior based upon. Do I have that right?

Reply:
One option was to use the following signals

- master replica count from install-config which is applied as unstructured to the `cluster-config-v1` configmap.
- require a second observation infrastructure as `BareMetal`. But since this appears to not be a clear signal I believe this option will not be 100% reasonable.

The next option which I feel will be more straight forward is that assisted-installer will embed a key to the `install-config` after it is rendered. This key will provide clear intent for example `installType: assisted`. To properly add this change we would need to add an optional field to the installConfig[1].

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-config-v1
  namespace: kube-system
data:
  install-config: |
    apiVersion: v1
    baseDomain: group-b.devcluster.openshift.com
    installType: assisted
    controlPlane:
      architecture: amd64
      hyperthreading: Enabled
      name: master
      platform:
        [..]
      replicas: 3
```

In which case the only gate for toleration is the existence of this key/value in `kube-system: cluster-config-v1` configmap.

[1] https://github.com/openshift/installer/blob/db45a7e83ade51ccbbbe92426488e2f680132b31/pkg/types/installconfig.go#L61

Reply:
This update to the design was discussed with @romfreiman but not updated to the draft which I am working on updating now.

Reply:
That seems to be aiming in a better direction. I think the `replicas: 2` approach has serious problems:

1. From a declarative intent standpoint is probably worse than weak, it's outright false: the user actually wants 3 masters, and the value of 2 is a hack that stands in conflict with actual intent, introducing complexities you can probably imagine but which we could get into if there's any question
2. The false declaration of intent will be indistinguishable from user error
3. Nothing else in the platform can be expected to support 2 master nodes via the scope of this proposal

I agree with you in principle that the general path forward is to get unambiguous intent from the user about the cluster topology from the installer. However, I'm not sure I have a persuasive argument for a new "installType" field which would be convincing to the installer folks and the rest of the component owners. Not only is it very generic,  it's actually coupled to a single specific (internal) initiative's concepts.

In the past when we've had similar issues (e.g. private clusters) we've arrived at a more [narrowly scoped piece of context](https://github.com/openshift/installer/commit/31a9d5274fee9e7dac64a784fdedc58775296587) supported by multiple use cases to inform the system and with cross-team buy-in (assuming there's the possibility that at some point in the future, some other non-etcd part of the platform may need to be aware of the context).

1. Is this toleration valid for all platform types, or just Bare Metal? Based on existing discussion on this EP, that assumption seems to be in question. If the toleration does in fact only apply to Bare Metal, could the toleration be expressed as a property of the Bare Metal platform type? If it can conceivably apply to other platforms (which seems true based on  https://github.com/openshift/enhancements/pull/48), a higher level field or inference seems more appropriate (but as I said, I'm not sure "install type" is it).
2. It's not clear to me how this use case is distinct from [Compact Clusters](https://github.com/openshift/enhancements/pull/48), but I'm unable to find any resulting additions to install config based on that work. What is the worker pool configuration or other state to indicate a compact cluster? For example, is it possible to infer this "compact" cluster topology based on an empty `compute` machine pool combined with a master pool size of 3? If so is it possible we don't need new installer context at all? cc @smarterclayton to make sure I'm not missing prior work here.

cc @abhinavdahiya @deads2k

Reply:
@derekwaynecarr and @deads2k gave me some background on compact clusters. My understanding is the compact topology is characterized by a bootstrap node and three masters. The assisted install topology does not fit that model precisely and represents a distinct installation topology. I think this answers my question (2) above. If my description of compact is correct, and if there's agreement that the assisted install topology is something new, my questions in (1) still stand.

Derek also had this interesting observation:
>The tension here is that this field is only meaningful on day 0, but has no relevance in day 2.
>I dont think we have defined a great communication channel for things that only matter on day 0 aside from the install config configmap we keep on the cluster as an initial signal for how day 0 was performed
>but we dont really want to use that as a source of truth for anything once the install completes

I agree but would also suggest day 0 has _some_ relevance on day 2: you can't get to day 2 without going through day 0. Ultimately we're saying that some specific vendor context (e.g. "we're installing via RH Assisted Installer") is required to bootstrap the cluster, and even implies a novel topology. The platform already needs awareness of some vendor platform context (e.g. AWS config), and it seems we're saying now the platform needs awareness of some _additional_ vendor context augmenting the existing platform info to make it to day 2.

Would a new field on the BareMetal platform make any sense? If we can't call it a new platform outright, it seems like we need _some_ new vendor specific fields.

How confident are we etcd is the only consumer of this additional context going forward?

Reply:
> Would a new field on the BareMetal platform make any sense? If we can't call it a new platform outright, it seems like we need _some_ new vendor specific fields.

It's not safe to assume that the assisted workflow is only used for actual bare metal clusters. There are plans under way already to expand the scope to use the assisted services to make it easier to deploy on platforms for which we don't have supported IPI. One of the reasons for using the "baremetal" platform now is to get the networking features, but at some point we hope to decouple those from the platform. So, if we need a new field, it shouldn't be tied to the platform type.

Reply:
I agree the condition was removed from consideration as a valid condition.